### PR TITLE
Separate connection specific args in EvaluateArgs.

### PR DESCRIPTION
**Pull request by ashithasantosh**
Adds nested class PerChannelArgs within EvaluateArgs. PerChannelArgs contains channel specific attributes (includes endpoint and auth_context fields).

This helps improve performance, by ensuring that the channel attributes are constructed only once, and not on a per-call basis. We will include EvaluateArgs::PerChannelArgs in authorization filter's channel_data structure.

This PR also updates the endpoint getters logic to support IPv6 format as well.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/25969)

### CODE REVIEW

**Review by markdroth — CHANGES_REQUESTED** [review]
Thanks for doing this!

My main comment is about the organization -- I don't think we need the per-channel and per-call data to be two different interfaces.

Please let me know if you have any questions.  Thanks!

> Reviewed 21 of 21 files at r1.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/25969#-:-MYGD_Sg5CUmb2hXpJ_3:bsknjkc)* status: all files reviewed, 6 unresolved discussions (waiting on @ashithasantosh, @jiangtaoli2016, and @ZhenLian)

___
*[src/core/lib/security/authorization/authorization_engine.h, line 71 at r1](https://reviewable.io/reviews/grpc/grpc/25969#-MYGBaRgEp9p9Zs-5Qf5:-MYGBaRh6YW33llnQn5K:bb865zs) ([raw file](https://github.com/grpc/grpc/blob/b132857b426d94bc10c8dc1a33bc1d1167606b90/src/core/lib/security/authorization/authorization_engine.h#L71)):*
> ```objc
> 
>   std::unique_ptr<mock_cel::Activation> CreateActivation(
>       const EvaluateChannelArgs& channel_args,
> ```

From the perspective of the engine, I don't think it should matter that some of the args are per-channel and others are per-call -- that should be an internal implementation detail of `EvaluateArgs`.

In other words, let's keep the `EvaluateArgs` API as it was before this PR, and simply change the implementation delegate to the new `EvaluateChannelArgs` object for the per-channel data.
___
*[src/core/lib/security/authorization/evaluate_channel_args.h, line 26 at r1](https://reviewable.io/reviews/grpc/grpc/25969#-MYGCWBj14nJYhWLtNb-:-MYGCWBj14nJYhWLtNb0:b-4s645o) ([raw file](https://github.com/grpc/grpc/blob/b132857b426d94bc10c8dc1a33bc1d1167606b90/src/core/lib/security/authorization/evaluate_channel_args.h#L26)):*
> ```objc
> 
> // Channel specific data.
> class EvaluateChannelArgs {
> ```

This class can be in the same file as the main `EvaluateArgs` class; there's no reason to have it in its own file.

In fact, this class can probably be nested inside of `EvaluateArgs`.  We can call it something like `EvaluateArgs::PerChannelData`.
___
*[src/core/lib/security/authorization/evaluate_channel_args.h, line 29 at r1](https://reviewable.io/reviews/grpc/grpc/25969#-MYFtETYB0rDspolC2K0:-MYFtETZBrtFaFgPLlbC:bhyrnlx) ([raw file](https://github.com/grpc/grpc/blob/b132857b426d94bc10c8dc1a33bc1d1167606b90/src/core/lib/security/authorization/evaluate_channel_args.h#L29)):*
> ```objc
>  public:
>   EvaluateChannelArgs(grpc_auth_context* auth_context, grpc_endpoint* endpoint);
>   ~EvaluateChannelArgs() = default;
> ```

This is not necessary.
___
*[src/core/lib/security/authorization/evaluate_channel_args.cc, line 50 at r1](https://reviewable.io/reviews/grpc/grpc/25969#-MYGAWwBAN2yOx73Nbyy:-MYGAWwCBzBJl27xAHws:b-6g5omc) ([raw file](https://github.com/grpc/grpc/blob/b132857b426d94bc10c8dc1a33bc1d1167606b90/src/core/lib/security/authorization/evaluate_channel_args.cc#L50)):*
> ```cpp
>   }
>   std::string port_str;
>   if (!SplitHostPort(uri->path(), address, &port_str)) {
> ```

I suggest using the variant of `SplitHostPort()` that returns `string_view`s, since that avoids unnecessarily allocating a string for the port.

```
absl::string_view address_view;
absl::string_view port_view;
if (!SplitHostPort(uri->path(), &address_view, &port_view)) {
  gpr_log(...);
  return;
}
*address = std::string(address_view);
```
___
*[src/core/lib/security/authorization/evaluate_channel_args.cc, line 69 at r1](https://reviewable.io/reviews/grpc/grpc/25969#-MYFtvQKDvwHES9BR9pf:-MYFtvQKDvwHES9BR9pg:bd0rb6x) ([raw file](https://github.com/grpc/grpc/blob/b132857b426d94bc10c8dc1a33bc1d1167606b90/src/core/lib/security/authorization/evaluate_channel_args.cc#L69)):*
> ```cpp
>                                          grpc_endpoint* endpoint) {
>   if (auth_context != nullptr) {
>     transport_security_type_ = std::string(GetAuthPropertyValue(
> ```

Instead of copying these strings, how about just having the `EvaluateChannelArgs` object hold a ref to the auth context, which contains the original strings?  That way, you can just store `string_view`s pointing inside of the auth context.
___
*[test/core/security/evaluate_channel_args_test.cc, line 26 at r1](https://reviewable.io/reviews/grpc/grpc/25969#-MYGD0RBCJPPtF5uvASo:-MYGD0RBCJPPtF5uvASp:b-k27y26) ([raw file](https://github.com/grpc/grpc/blob/b132857b426d94bc10c8dc1a33bc1d1167606b90/test/core/security/evaluate_channel_args_test.cc#L26)):*
> ```cpp
> namespace grpc_core {
> 
> TEST(EvaluateChannelArgsTest, HandlesNullEndpoint) {
> ```

All of these tests can go in a single file, both per-channel and per-call data.

**Review by ashithasantosh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/25969)* status: 6 of 21 files reviewed, 6 unresolved discussions (waiting on @ashithasantosh, @jiangtaoli2016, @markdroth, and @ZhenLian)

___
*[src/core/lib/security/authorization/authorization_engine.h, line 71 at r1](https://reviewable.io/reviews/grpc/grpc/25969#-MYGBaRgEp9p9Zs-5Qf5:-MYIrQIT0KJgg0kaRWwV:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/b132857b426d94bc10c8dc1a33bc1d1167606b90/src/core/lib/security/authorization/authorization_engine.h#L71)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

From the perspective of the engine, I don't think it should matter that some of the args are per-channel and others are per-call -- that should be an internal implementation detail of `EvaluateArgs`.

In other words, let's keep the `EvaluateArgs` API as it was before this PR, and simply change the implementation delegate to the new `EvaluateChannelArgs` object for the per-channel data.
</blockquote></details>

Done.
___
*[src/core/lib/security/authorization/evaluate_channel_args.h, line 26 at r1](https://reviewable.io/reviews/grpc/grpc/25969#-MYGCWBj14nJYhWLtNb-:-MYJH6if97WuuK8Ld_cL:b-9jxbnf) ([raw file](https://github.com/grpc/grpc/blob/b132857b426d94bc10c8dc1a33bc1d1167606b90/src/core/lib/security/authorization/evaluate_channel_args.h#L26)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This class can be in the same file as the main `EvaluateArgs` class; there's no reason to have it in its own file.

In fact, this class can probably be nested inside of `EvaluateArgs`.  We can call it something like `EvaluateArgs::PerChannelData`.
</blockquote></details>

I added PerChannelArgs struct within EvaluateArgs. PTAL.
We can then include PerChannelArgs in channel_data (of rbac filter).
___
*[src/core/lib/security/authorization/evaluate_channel_args.h, line 29 at r1](https://reviewable.io/reviews/grpc/grpc/25969#-MYFtETYB0rDspolC2K0:-MYIrU2Z7O-DPO2Dm9mr:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/b132857b426d94bc10c8dc1a33bc1d1167606b90/src/core/lib/security/authorization/evaluate_channel_args.h#L29)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This is not necessary.
</blockquote></details>

Done.
___
*[src/core/lib/security/authorization/evaluate_channel_args.cc, line 50 at r1](https://reviewable.io/reviews/grpc/grpc/25969#-MYGAWwBAN2yOx73Nbyy:-MYIrYuAEMXPlJh92NBD:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/b132857b426d94bc10c8dc1a33bc1d1167606b90/src/core/lib/security/authorization/evaluate_channel_args.cc#L50)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I suggest using the variant of `SplitHostPort()` that returns `string_view`s, since that avoids unnecessarily allocating a string for the port.

```
absl::string_view address_view;
absl::string_view port_view;
if (!SplitHostPort(uri->path(), &address_view, &port_view)) {
  gpr_log(...);
  return;
}
*address = std::string(address_view);
```
</blockquote></details>

Done.
___
*[test/core/security/evaluate_channel_args_test.cc, line 26 at r1](https://reviewable.io/reviews/grpc/grpc/25969#-MYGD0RBCJPPtF5uvASo:-MYIr_cTDqRuUoAY2N-6:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/b132857b426d94bc10c8dc1a33bc1d1167606b90/test/core/security/evaluate_channel_args_test.cc#L26)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

All of these tests can go in a single file, both per-channel and per-call data.
</blockquote></details>

Done.

**Review by markdroth — CHANGES_REQUESTED** [review]
This looks great!  Just a few minor comments remaining.

Please let me know if you have any questions.  Thanks!

> Reviewed 1 of 21 files at r1, 15 of 15 files at r2.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/25969#-:-MYL7twoDgGxEUbhxz6k:b-nmtjq6)* status: all files reviewed, 3 unresolved discussions (waiting on @ashithasantosh, @jiangtaoli2016, and @ZhenLian)

___
*[test/core/security/evaluate_args_test.cc, line 26 at r2](https://reviewable.io/reviews/grpc/grpc/25969#-MYL6VShD1YLscnQUHgP:-MYL6VShD1YLscnQUHgQ:b-bol98r) ([raw file](https://github.com/grpc/grpc/blob/188e117fa03383a4dbb0345a0c7a9db43215905b/test/core/security/evaluate_args_test.cc#L26)):*
> ```cpp
> namespace grpc_core {
> 
> TEST(EvaluateArgsTest, EmptyMetadata) {
> ```

I think these tests would benefit from the same framework that you added in authorization_matchers_test.cc #25653.  How about adding that framework as a common library, to be used in both sets of tests?
___
*[src/core/lib/security/authorization/evaluate_args.h, line 35 at r2](https://reviewable.io/reviews/grpc/grpc/25969#-MYL4sElDTYFumGy9jsx:-MYL4sElDTYFumGy9jsy:bxvd7ip) ([raw file](https://github.com/grpc/grpc/blob/188e117fa03383a4dbb0345a0c7a9db43215905b/src/core/lib/security/authorization/evaluate_args.h#L35)):*
> ```objc
>     PerChannelArgs(grpc_auth_context* auth_context, grpc_endpoint* endpoint);
> 
>     grpc_core::RefCountedPtr<grpc_auth_context> auth_ctx;
> ```

If this is going to be a struct instead of a class, with all data members being public, then let's not hold the ref to the auth context here.  Instead, let's hold this ref in the filter's channel data object, which will also own the `PerChannelData` struct.  That will ensure that they have the same lifetime.

Please add a comment on the `PerChannelArgs` ctor that the caller is responsible for ensuring that the auth context lives until after the `PerChannelArgs` struct is no longer used.
___
*[src/core/lib/security/authorization/evaluate_args.h, line 46 at r2](https://reviewable.io/reviews/grpc/grpc/25969#-MYL4GJo9pb7qtxoB0Vy:-MYL4GJo9pb7qtxoB0Vz:b-y7fke) ([raw file](https://github.com/grpc/grpc/blob/188e117fa03383a4dbb0345a0c7a9db43215905b/src/core/lib/security/authorization/evaluate_args.h#L46)):*
> ```objc
> 
>   EvaluateArgs(grpc_metadata_batch* metadata,
>                std::unique_ptr<PerChannelArgs> channel_args)
> ```

I don't think we want this to be a `unique_ptr<>`; we want it to be a raw pointer, and `EvaluateArgs` will *not* take ownership of it.  The reason for this is that the filter's channel data object will maintain ownership of the `PerChannelArgs` struct,  and we will create a separate `EvaluateArgs` for each call data object.  We don't want to have to copy the `PerChannelArgs` struct for each call; instead, we want to just use the existing copy from the channel data object.

**Review by ashithasantosh — COMMENTED** [review]
Thanks for the review, Mark! I resolved all comments. PTAL.

> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/25969#-:-MYM1F1cDxXpxleqoFgI:b44h87t)* status: 15 of 22 files reviewed, 3 unresolved discussions (waiting on @jiangtaoli2016, @markdroth, and @ZhenLian)

___
*[test/core/security/evaluate_args_test.cc, line 26 at r2](https://reviewable.io/reviews/grpc/grpc/25969#-MYL6VShD1YLscnQUHgP:-MYM1CS6ErblVTlRFOvi:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/188e117fa03383a4dbb0345a0c7a9db43215905b/test/core/security/evaluate_args_test.cc#L26)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I think these tests would benefit from the same framework that you added in authorization_matchers_test.cc #25653.  How about adding that framework as a common library, to be used in both sets of tests?
</blockquote></details>

Done.
___
*[src/core/lib/security/authorization/evaluate_args.h, line 35 at r2](https://reviewable.io/reviews/grpc/grpc/25969#-MYL4sElDTYFumGy9jsx:-MYM115iC9YbE7qF3Qqe:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/188e117fa03383a4dbb0345a0c7a9db43215905b/src/core/lib/security/authorization/evaluate_args.h#L35)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

If this is going to be a struct instead of a class, with all data members being public, then let's not hold the ref to the auth context here.  Instead, let's hold this ref in the filter's channel data object, which will also own the `PerChannelData` struct.  That will ensure that they have the same lifetime.

Please add a comment on the `PerChannelArgs` ctor that the caller is responsible for ensuring that the auth context lives until after the `PerChannelArgs` struct is no longer used.
</blockquote></details>

Done.
___
*[src/core/lib/security/authorization/evaluate_args.h, line 46 at r2](https://reviewable.io/reviews/grpc/grpc/25969#-MYL4GJo9pb7qtxoB0Vy:-MYM11je0wcEFeQoEueu:b-azgjw9) ([raw file](https://github.com/grpc/grpc/blob/188e117fa03383a4dbb0345a0c7a9db43215905b/src/core/lib/security/authorization/evaluate_args.h#L46)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I don't think we want this to be a `unique_ptr<>`; we want it to be a raw pointer, and `EvaluateArgs` will *not* take ownership of it.  The reason for this is that the filter's channel data object will maintain ownership of the `PerChannelArgs` struct,  and we will create a separate `EvaluateArgs` for each call data object.  We don't want to have to copy the `PerChannelArgs` struct for each call; instead, we want to just use the existing copy from the channel data object.
</blockquote></details>

That's right! Thanks! Updated to use raw pointer.

**Review by markdroth — CHANGES_REQUESTED** [review]
This looks great!  I have a couple of minor comments, but I'll go ahead and approve.  Feel free to merge once you've resolved those comments.

Please let me know if you have any questions.  Thanks!

> Reviewed 7 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/25969#-:-MYM5abl-4g0Ybzt1cWE:b5lg546)* status: all files reviewed, 2 unresolved discussions (waiting on @ashithasantosh, @jiangtaoli2016, and @ZhenLian)

___
*[test/core/security/evaluate_args_test.cc, line 26 at r2](https://reviewable.io/reviews/grpc/grpc/25969#-MYL6VShD1YLscnQUHgP:-MYM4h7r6eP3b7u0ghM9:bzfswjq) ([raw file](https://github.com/grpc/grpc/blob/188e117fa03383a4dbb0345a0c7a9db43215905b/test/core/security/evaluate_args_test.cc#L26)):*
<details><summary><i>Previously, ashithasantosh (Ashitha Santhosh) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks, that looks good.  I suggest still having a test class here, just to avoid having to declare the `AuthorizationArgs` object in every individual test.

```
class EvaluateArgsTest : public ::testing::Test {
 protected:
  AuthorizationArgs args_;
};
```

That way, individual tests can just interact with `args_` instead of declaring it themselves.
___
*[test/core/util/authorization_args.h, line 29 at r3](https://reviewable.io/reviews/grpc/grpc/25969#-MYM4AHC-YZ4Dr85-Zlq:-MYM4AHC-YZ4Dr85-Zlr:b9v7q8a) ([raw file](https://github.com/grpc/grpc/blob/73f07590073f5db5d5c84ce04f8c8aeb211d1abf/test/core/util/authorization_args.h#L29)):*
> ```objc
> namespace grpc_core {
> 
> class AuthorizationArgs {
> ```

Suggest calling this something like `FakeEvaluateArgs` or `EvaluateArgsTestUtil` (and rename the file accordingly).

**Review by markdroth — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/25969#-:-MYM5mHt9VdECfc1iU7C:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @ashithasantosh, @jiangtaoli2016, and @ZhenLian)

**Review by ashithasantosh — COMMENTED** [review]
Thanks for the quick review, Mark!:)

> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/25969#-:-MYMHT272sU0ItXvPDUi:b-1dt050)* status: 17 of 22 files reviewed, 2 unresolved discussions (waiting on @jiangtaoli2016, @markdroth, and @ZhenLian)

___
*[test/core/security/evaluate_args_test.cc, line 26 at r2](https://reviewable.io/reviews/grpc/grpc/25969#-MYL6VShD1YLscnQUHgP:-MYMHXjA9BQC5Gr41-Dp:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/188e117fa03383a4dbb0345a0c7a9db43215905b/test/core/security/evaluate_args_test.cc#L26)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Thanks, that looks good.  I suggest still having a test class here, just to avoid having to declare the `AuthorizationArgs` object in every individual test.

```
class EvaluateArgsTest : public ::testing::Test {
 protected:
  AuthorizationArgs args_;
};
```

That way, individual tests can just interact with `args_` instead of declaring it themselves.
</blockquote></details>

Done.
___
*[test/core/util/authorization_args.h, line 29 at r3](https://reviewable.io/reviews/grpc/grpc/25969#-MYM4AHC-YZ4Dr85-Zlq:-MYMHYRY7NZS6IC9JJLC:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/73f07590073f5db5d5c84ce04f8c8aeb211d1abf/test/core/util/authorization_args.h#L29)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Suggest calling this something like `FakeEvaluateArgs` or `EvaluateArgsTestUtil` (and rename the file accordingly).
</blockquote></details>

Done.

**Review by ZhenLian — APPROVED** [review]
Thanks Mark for the detailed review and Ashitha for the PR! I don't see anything worth commenting, so I will approve right away.
I noticed that the there are more-than-normal failing presubmit tests. If you are sure they are irrelevant, feel free to merge :)

**Comment by ashithasantosh** [comment]
Thanks for the review, Zhen!:) 
Majority of the test failures were caused by tool failure. On rerunning these tests, we see two failures, which are irrelevant.

File: src/core/lib/security/authorization/evaluate_args.h
Comment (jiangtaoli2016) [inline src/core/lib/security/authorization/evaluate_args.h:38]: DNS name will be plumbed in a different PR?

Comment (ashithasantosh) [reply src/core/lib/security/authorization/evaluate_args.h:38]: DNS name plumbing PR is already submitted. Since I submitted that today, after this PR was approved, I thought I will add it in a follow-up PR.

**Comment by ashithasantosh** [comment]
Thanks for the review, Jiangtao!:)

**Review by jiangtaoli2016 — APPROVED** [review]

**Pull request merged**
